src/cpp/_pythia8.cpp

```diff
@@ -103,6 +105,73 @@ py::array_t<int> event_array_children(Event &event)
     return py::array_t<int>(shape, strides, ptr1);
 }
 
+// refills "event" stack with particles
+void fill_event(Event &event,
+                    py::array_t<int> &pid,
+                    py::array_t<int> &status,
+                    py::array_t<double> &px,
+                    py::array_t<double> &py,
+                    py::array_t<double> &pz,
+                    py::array_t<double> &energy,
+                    py::array_t<double> &mass) 
+{
+    const double spare_mem = 1.5L;
```

Comment:
If you use reserve, you should not use spare_mem, I think. Unlike for vector::push_back, you won't get O(1) asymptotic behavior this way.

Reply:
> Unlike for vector::push_back, you won't get O(1) asymptotic behavior this way.

I didn't get what you mean. But thanks to this comment I realized that I need to reserve  `entry` 2x larger of event, because it is used to save initial entry + decayed particles. I checked that for the default settings the array usually increase 1.5 - 1.8 times. If all particles decay, then the size increases 5x. I guess 2x is conservative estimation.

Reply:
Ah, I see now what this factor 1.5-1.8 was for. I thought you try to imitate how std::vector grows.

Anyway, you are overoptimising this. Pythia is not shrinking the particle vector in the event. So all this optimisation of trying to reserve enough space for all particles will only run once or twice until the particle vector has grown large enough to store all particles. Afterwards, it will just keep its size.

I think you should remove all this code to reserve space in the vector and just call the public API. Use Event::reset() and Event::append(). This is going to be efficient enough.

Reply:
Ok. I leave a bare minimum version of the function. I will change when I am sure that the optimization with reserve works.